modules/grandpa/src/lib.rs

```diff
@@ -134,16 +163,18 @@ pub mod pallet {
 			// "travelling back in time" (which could be indicative of something bad, e.g a hard-fork).
 			ensure!(best_finalized.number() < number, <Error<T, I>>::OldHeader);
 
-			verify_justification::<T, I>(&justification, hash, *number)?;
+			let authority_set = <CurrentAuthoritySet<T, I>>::get();
+			let set_id = authority_set.set_id;
+			let (precommits, votes) = verify_justification::<T, I>(&justification, hash, *number, authority_set)?;
 
-			try_enact_authority_change::<T, I>(&finality_target)?;
+			let _enacted = try_enact_authority_change::<T, I>(&finality_target, set_id)?;
 			<BestFinalized<T, I>>::put(hash);
 			<ImportedHeaders<T, I>>::insert(hash, finality_target);
 			<RequestCount<T, I>>::mutate(|count| *count += 1);
 
 			log::info!(target: "runtime::bridge-grandpa", "Succesfully imported finalized header with hash {:?}!", hash);
 
-			Ok(().into())
+			Ok(Some(T::WeightInfo::submit_finality_proof(votes as u32, precommits as u32)).into())
```

Comment:
I think this is incorrect and may result in an even higher weight than declared (BTW, what happens in such case?).

In the benchmarks we have two parameters: `forks` and `depth` which correspond to number of pre-commits and length of forks the pre-commits are on - i.e. the total number of headers in justification will be `headers = forks * depth`. Here, afaiu, `votes` indicates the total number of headers in the justification, so the weight we calculate is going to be for a worst case, where we have `votes * precommits` number of headers.

Reply:
I don't quite follow your math here. The benchmarks are set up to account for the effects of two things: how long it takes to check pre-commits (`p`), and how long it takes to check vote ancestries (`s`). They are used in the weight formula as follows:

```rust
(0 as Weight)
	.saturating_add((3_248_661_000 as Weight).saturating_mul(s as Weight))
	.saturating_add((776_552_000 as Weight).saturating_mul(p as Weight))
```

As you can see these two things can be thought of as independent of one another.

Maybe can you clarify what you mean by "total number of headers in the justification"? Because I don't see how we can get a higher than declared weight by using `votes`.

> BTW, what happens in such case?

I don't think the benchmarks are doing to charge your account more than the declared weight, so I guess it would open up a DoS vector for the chain.

Reply:
>  how long it takes to check vote ancestries (s). 

That's not accurate. `s` is passed as `depth` to `make_justification_for_header` function. Take a look at the code that [generates justifications here](https://github.com/paritytech/parity-bridges-common/blob/a937b074f76a36868e358d222c133318e6087cb5/primitives/test-utils/src/lib.rs#L106).
For every `fork` you create a chain of `depth` blocks and add all of them to `votes_ancestries`. Hence `votes_ancestries ~ forks * depth`.

So the weight calculation is based on the worst case and the interpretation of `s` and `p` should be "maximal length of EACH fork" and "number of forks" accordingly.

Hence if you pass `votes` coming from actual justification you end up over estimating the weight quite significantly - I think you should rather be be passing "average length of EACH fork" calculated as `p = pre_commits.len(); s = (votes.len() + p - 1) / p` to get realistic results.

Reply:
Okay I think I see what you mean. Basically, our weight formula is a function of `s` and `p`:

```rust
fn weight(s, p) -> Weight;
```

During the refund, we call `weight(votes, precommits)`, which can be thought of as `weight(s * p, p)`, which is clearly an overestimate of the declared weight.

Since what we really want is `weight(s, p)` you're suggesting that we take the average in order to isolate `s` from `s * p`. 

This makes sense to me now, thanks for spotting this bug!